File: sdk/c/src/providers/mailmomy.c
/**
 * mailmomy.com 渠道
 *
 * 免注册、无鉴权、无 CAPTCHA 的纯 GET JSON 临时邮箱，token 即邮箱地址本身。
 *
 * 域名池（可选拉取）:
 *   GET https://mailmomy.com/api/domains/active
 *   返回 JSON 字符串数组，如 ["mailmomy.com","282mail.com",...]
 *   从中随机选一个域名；请求失败或为空时回退到默认 "mailmomy.com"。
 *
 * 创建邮箱:
 *   本地随机 10 位 [a-z0-9] local part，email = local@<域名>，token = email。
 *
 * 获取邮件:
 *   GET https://mailmomy.com/api/mail/messages?to=<email>&page=1&limit=20
 *   返回:
 * {"emails":[{"id","recipient","from","subject","message","bodyText","receivedAt"}],
 * ...}
 */
#include "tempmail_internal.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAILMOMY_BASE "https://mailmomy.com"
#define MAILMOMY_DEFAULT_DOMAIN "mailmomy.com"
#define MAILMOMY_MAX_MAILS 128

/* 公共 UA */
static const char *mailmomy_ua =
    "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/149.0.0.0 Safari/537.36";

/* URL 编码单个字符（RFC3986 未保留字符原样保留，其余百分号编码） */
static int mailmomy_url_encode_char(char c, char *out) {
  static const char hex[] = "0123456789ABCDEF";
  unsigned char uc = (unsigned char)c;
  if ((uc >= 'A' && uc <= 'Z') || (uc >= 'a' && uc <= 'z') ||
      (uc >= '0' && uc <= '9') || uc == '-' || uc == '_' || uc == '.' ||
      uc == '~') {
    out[0] = c;
    return 1;
  }
  out[0] = '%';
  out[1] = hex[(uc >> 4) & 0x0F];
  out[2] = hex[uc & 0x0F];
  return 3;
}

/* URL 编码字符串（堆分配，调用者 free） */
static char *mailmomy_url_encode(const char *s) {
  if (!s)
    return NULL;
  size_t len = strlen(s);
  char *out = (char *)malloc(len * 3 + 1);
  if (!out)
    return NULL;
  size_t o = 0;
  for (size_t i = 0; i < len; i++) {
    o += (size_t)mailmomy_url_encode_char(s[i], out + o);
  }
  out[o] = '\0';
  return out;
}

/* 随机生成邮箱前缀（小写字母数字） */
static void mailmomy_random_local(char *out, size_t cap) {
  static const char chars[] = "abcdefghijklmnopqrstuvwxyz0123456789";
  if (!out || cap < 2)
    return;
  size_t n = 10;
  if (n >= cap)
    n = cap - 1;
  for (size_t i = 0; i < n; i++) {
    out[i] = chars[rand() % (int)(sizeof(chars) - 1)];
  }
  out[n] = '\0';
}

/**
 * 拉取域名池并随机挑选一个域名（堆分配，调用者 free）。
 * 请求失败、非数组或为空时回退到默认 "mailmomy.com"。
 */
static char *mailmomy_pick_domain(int timeout) {
  const char *hdrs[] = {mailmomy_ua, "Accept: application/json", NULL};

  tm_http_response_t *resp = tm_http_request(
      TM_HTTP_GET, MAILMOMY_BASE "/api/domains/active", hdrs, NULL, timeout);
  if (!resp || resp->status != 200 || !resp->body) {
    TM_LOG_WARN("[mailmomy] 获取域名列表失败, status=%ld，回退默认域名",
                resp ? resp->status : 0);
    tm_http_response_free(resp);
    return tm_strdup(MAILMOMY_DEFAULT_DOMAIN);
  }

  cJSON *json = cJSON_Parse(resp->body);
  tm_http_response_free(resp);
  if (!json || !cJSON_IsArray(json)) {
    TM_LOG_WARN("[mailmomy] 域名列表非数组，回退默认域名");
    if (json)
      cJSON_Delete(json);
    return tm_strdup(MAILMOMY_DEFAULT_DOMAIN);
  }

  int n = cJSON_GetArraySize(json);
  if (n <= 0) {
    cJSON_Delete(json);
    return tm_strdup(MAILMOMY_DEFAULT_DOMAIN);
  }

  /* 收集有效（非空字符串）域名下标 */
  int *pool = (int *)malloc((size_t)n * sizeof(int));
  if (!pool) {
    cJSON_Delete(json);
    return tm_strdup(MAILMOMY_DEFAULT_DOMAIN);
  }

  int pool_n = 0;
  for (int i = 0; i < n; i++) {
    const cJSON *d = cJSON_GetArrayItem(json, i);
    if (cJSON_IsString(d) && d->valuestring && d->valuestring[0]) {
      pool[pool_n++] = i;
    }
  }

  if (pool_n == 0) {
    free(pool);
    cJSON_Delete(json);
    return tm_strdup(MAILMOMY_DEFAULT_DOMAIN);
  }

  int picked = pool[rand() % pool_n];
  free(pool);

  const cJSON *chosen = cJSON_GetArrayItem(json, picked);
  char *domain = tm_strdup(chosen->valuestring);
  cJSON_Delete(json);
  return domain ? domain : tm_strdup(MAILMOMY_DEFAULT_DOMAIN);
}

/**
 * 创建 mailmomy.com 临时邮箱
 */
tm_email_info_t *tm_provider_mailmomy_generate(void) {
  int timeout =
      tm_get_config()->timeout_secs > 0 ? tm_get_config()->timeout_secs : 15;

  char *domain = mailmomy_pick_domain(timeout);
  if (!domain)
    return NULL;

  char local[16];
  mailmomy_random_local(local, sizeof(local));

  size_t email_cap = strlen(local) + strlen(domain) + 2;
  char *email_addr = (char *)malloc(email_cap);
  if (!email_addr) {
    free(domain);
    return NULL;
  }
  snprintf(email_addr, email_cap, "%s@%s", local, domain);
  free(domain);

  tm_email_info_t *info = tm_email_info_new();
  if (!info) {
    free(email_addr);
    return NULL;
  }
  info->channel = CHANNEL_MAILMOMY;
  info->email = email_addr;
  info->token = tm_strdup(email_addr); /* token 即邮箱地址 */
  if (!info->token) {
    tm_free_email_info(info);
    return NULL;
  }
  info->expires_at = 0;
  info->created_at = tm_strdup("");

  TM_LOG_DBG("[mailmomy] 创建邮箱成功: %s", info->email);
  return info;
}

/**
 * 获取 mailmomy.com 邮件列表
 */
tm_email_t *tm_provider_mailmomy_get_emails(const char *token,
                                            const char *email, int *count) {
  *count = -1;
  const char *addr = (token && token[0]) ? token : email;
  if (!addr || !addr[0]) {
    TM_LOG_WARN("[mailmomy] 缺少邮箱地址");
    return NULL;
  }

  int timeout =
      tm_get_config()->timeout_secs > 0 ? tm_get_config()->timeout_secs : 15;

  char *enc_addr = mailmomy_url_encode(addr);
  if (!enc_addr)
    return NULL;

  size_t url_cap = strlen(MAILMOMY_BASE) + strlen(enc_addr) + 64;
  char *url = (char *)malloc(url_cap);
  if (!url) {
    free(enc_addr);
    return NULL;
  }
  snprintf(url, url_cap,
           MAILMOMY_BASE "/api/mail/messages?to=%s&page=1&limit=20", enc_addr);
  free(enc_addr);

  const char *hdrs[] = {mailmomy_ua, "Accept: application/json", NULL};

  tm_http_response_t *resp =
      tm_http_request(TM_HTTP_GET, url, hdrs, NULL, timeout);
  free(url);
  if (!resp || resp->status != 200 || !resp->body) {
    TM_LOG_ERR("[mailmomy] 获取邮件失败, status=%ld", resp ? resp->status : 0);
    tm_http_response_free(resp);
    return NULL;
  }

  cJSON *json = cJSON_Parse(resp->body);
  tm_http_response_free(resp);
  if (!json) {
    TM_LOG_ERR("[mailmomy] 邮件响应解析失败");
    return NULL;
  }

  const cJSON *j_emails = cJSON_GetObjectItemCaseSensitive(json, "emails");
  if (!cJSON_IsArray(j_emails)) {
    cJSON_Delete(json);
    *count = 0;
    return NULL;
  }

  int n = cJSON_GetArraySize(j_emails);
  if (n == 0) {
    cJSON_Delete(json);
    *count = 0;
    return NULL;
  }
  if (n > MAILMOMY_MAX_MAILS)
    n = MAILMOMY_MAX_MAILS;

  tm_email_t *emails = tm_emails_new(n);
  if (!emails) {
    cJSON_Delete(json);
    return NULL;
  }

  int valid = 0;
  for (int i = 0; i < n; i++) {
    const cJSON *msg = cJSON_GetArrayItem(j_emails, i);
    if (!cJSON_IsObject(msg))
      continue;

    /*
     * mailmomy 字段名与 normalize 候选 key 不完全一致：
     *   message -> html，bodyText -> text（bodyText 缺失时退回 message 作正文）
     * 构造一个规整对象供 tm_normalize_email 消费；id/from/subject/receivedAt
     * 直接沿用。
     */
    const cJSON *j_id = cJSON_GetObjectItemCaseSensitive(msg, "id");
    const cJSON *j_from = cJSON_GetObjectItemCaseSensitive(msg, "from");
    const cJSON *j_recipient =
        cJSON_GetObjectItemCaseSensitive(msg, "recipient");
    const cJSON *j_subject = cJSON_GetObjectItemCaseSensitive(msg, "subject");
    const cJSON *j_message = cJSON_GetObjectItemCaseSensitive(msg, "message");
    const cJSON *j_bodytext = cJSON_GetObjectItemCaseSensitive(msg, "bodyText");
    const cJSON *j_received =
        cJSON_GetObjectItemCaseSensitive(msg, "receivedAt");

    cJSON *mapped = cJSON_CreateObject();
    if (!mapped)
      continue;
    if (cJSON_IsString(j_id) && j_id->valuestring)
      cJSON_AddStringToObject(mapped, "id", j_id->valuestring);
    if (cJSON_IsString(j_from) && j_from->valuestring)
      cJSON_AddStringToObject(mapped, "from", j_from->valuestring);
    if (cJSON_IsString(j_recipient) && j_recipient->valuestring)
      cJSON_AddStringToObject(mapped, "to", j_recipient->valuestring);
    if (cJSON_IsString(j_subject) && j_subject->valuestring)
      cJSON_AddStringToObject(mapped, "subject", j_subject->valuestring);
    if (cJSON_IsString(j_message) && j_message->valuestring)
      cJSON_AddStringToObject(mapped, "html", j_message->valuestring);
    if (cJSON_IsString(j_bodytext) && j_bodytext->valuestring &&
        j_bodytext->valuestring[0])
      cJSON_AddStringToObject(mapped, "text", j_bodytext->valuestring);
    else if (cJSON_IsString(j_message) && j_message->valuestring)
      cJSON_AddStringToObject(mapped, "text", j_message->valuestring);
    if (cJSON_IsString(j_received) && j_received->valuestring)
      cJSON_AddStringToObject(mapped, "receivedAt", j_received->valuestring);

    emails[valid] = tm_normalize_email(mapped, addr);
    cJSON_Delete(mapped);
    valid++;
  }

  cJSON_Delete(json);

  if (valid == 0) {
    free(emails);
    *count = 0;
    return NULL;
  }

  *count = valid;
  return emails;
}

File: sdk/rust/src/providers/mailmomy.rs
/*!
 * mailmomy 渠道实现
 * 网站: https://mailmomy.com
 *
 * 独立自研后端，免注册、无鉴权、无 CAPTCHA、纯 GET JSON API：
 *   1. GET /api/domains/active 拉取可用域名池（JSON 字符串数组），随机选一个域名
 *   2. 本地随机 10 位 [a-z0-9] local part 组成邮箱地址
 *   3. GET /api/mail/messages?to={email}&page=1&limit=20 读取邮件列表
 *
 * token 即邮箱地址本身。
 */

use rand::Rng;
use serde_json::Value;
use wreq::Client;

use crate::config::{block_on, get_current_ua, http_client};
use crate::normalize::normalize_email;
use crate::types::{Channel, Email, EmailInfo};

const BASE_URL: &str = "https://mailmomy.com";
const DEFAULT_DOMAIN: &str = "mailmomy.com";

/// 生成随机本地部分（小写字母 + 数字）
fn random_local(length: usize) -> String {
    let chars: Vec<char> = "abcdefghijklmnopqrstuvwxyz0123456789".chars().collect();
    let mut rng = rand::thread_rng();
    (0..length)
        .map(|_| chars[rng.gen_range(0..chars.len())])
        .collect()
}

/// URL 编码邮箱地址中的 @ 符号
fn encode_mail(email: &str) -> String {
    email.replace('@', "%40")
}

/// 拉取 mailmomy 当前可用域名池并随机选择一个
/// 接口失败或返回空时回退到默认 mailmomy.com
async fn pick_domain(client: &Client) -> String {
    let resp = client
        .get(format!("{}/api/domains/active", BASE_URL))
        .header("User-Agent", get_current_ua())
        .header("Accept", "application/json")
        .send()
        .await;

    let resp = match resp {
        Ok(r) if r.status().is_success() => r,
        _ => return DEFAULT_DOMAIN.to_string(),
    };

    let data: Value = match resp.json().await {
        Ok(v) => v,
        Err(_) => return DEFAULT_DOMAIN.to_string(),
    };

    let list: Vec<String> = match data.as_array() {
        Some(arr) => arr
            .iter()
            .filter_map(|v| v.as_str())
            .filter(|s| !s.is_empty())
            .map(|s| s.to_string())
            .collect(),
        None => return DEFAULT_DOMAIN.to_string(),
    };

    if list.is_empty() {
        return DEFAULT_DOMAIN.to_string();
    }
    let idx = rand::thread_rng().gen_range(0..list.len());
    list[idx].clone()
}

/// 创建 mailmomy.com 临时邮箱
/// 服务免注册，直接构造 <local>@<域名> 即可收信
pub fn generate_email() -> Result<EmailInfo, String> {
    let client = http_client();

    block_on(async {
        let domain = pick_domain(&client).await;
        let email = format!("{}@{}", random_local(10), domain);
        Ok(EmailInfo {
            channel: Channel::Mailmomy,
            email: email.clone(),
            token: Some(email),
            expires_at: None,
            created_at: None,
        })
    })
}

/// 获取 mailmomy.com 邮件列表
/// GET /api/mail/messages?to={email}&page=1&limit=20
/// 返回 {emails:[{id,recipient,from,subject,message,bodyText,receivedAt}], total, page, limit, pages}
pub fn get_emails(email: &str) -> Result<Vec<Email>, String> {
    if email.is_empty() {
        return Err("mailmomy: 缺少邮箱地址".into());
    }
    let client = http_client();

    block_on(async {
        let url = format!(
            "{}/api/mail/messages?to={}&page=1&limit=20",
            BASE_URL,
            encode_mail(email)
        );
        let resp = client
            .get(url)
            .header("User-Agent", get_current_ua())
            .header("Accept", "application/json")
            .send()
            .await
            .map_err(|e| format!("mailmomy: 获取邮件请求失败: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!("mailmomy: 获取邮件失败 HTTP {}", resp.status()));
        }

        let data: Value = resp
            .json()
            .await
            .map_err(|e| format!("mailmomy: 解析邮件响应失败: {}", e))?;

        let list = match data.get("emails").and_then(|v| v.as_array()) {
            Some(arr) if !arr.is_empty() => arr,
            _ => return Ok(vec![]),
        };

        let mut out = Vec::with_capacity(list.len());
        for msg in list {
            let recipient = json_str(msg.get("recipient"));
            let to = if recipient.is_empty() {
                email.to_string()
            } else {
                recipient
            };
            let message = json_str(msg.get("message"));
            let body_text = json_str(msg.get("bodyText"));
            let text = if body_text.is_empty() {
                message.clone()
            } else {
                body_text
            };
            let flat = serde_json::json!({
                "id": json_str(msg.get("id")),
                "from": json_str(msg.get("from")),
                "to": to,
                "subject": json_str(msg.get("subject")),
                "text": text,
                "html": message,
                "date": json_str(msg.get("receivedAt")),
                "isRead": false,
            });
            out.push(normalize_email(&flat, email));
        }
        Ok(out)
    })
}

/// 安全提取 JSON 值为字符串（字符串原样、数字转文本）
fn json_str(v: Option<&Value>) -> String {
    match v {
        Some(v) if v.is_string() => v.as_str().unwrap_or("").to_string(),
        Some(v) if v.is_number() => v.to_string(),
        _ => String::new(),
    }
}
